### ISSUE TITLE
Certificates Provider/Resolver

### ISSUE THREAD

**Issue by ldez** [containous/traefik#4872]
### Do you want to request a *feature* or report a *bug*?

Feature

### What did you expect to see?

- Dynamic wildcard
- Challenge by domain(s) / multiple challenges
- DNS provider by domain(s) / multiple DNS providers
- Multiple certificates providers (acme, vault, ...)

---
### Static

```toml
[certificatesResolvers.default.acme]
  tlsStore = "test"
  storage = "acme.json"
  caServer = "https://acme-staging-v02.api.letsencrypt.org/directory"
  email = "test@test.com"
  keyType = "RSA4096"
  acmeLogging = true
#  resolvers = ["1.1.1.1:53", "8.8.8.8:53"]

  [certificatesResolvers.default.acme.TLSchallenge]
```

### Dynamic

```toml
[http.routers.router1]
  service = "service2"
  rule = "Host(`goo.snitest.com`)"
  [http.routers.router1.tls]
#    store = "test"
    certResolver = "default"
```

## Case: wildcard

Note: If a section `domain` is defined => it doesn't use the Host rules.

### Static

```toml
[certificatesResolvers.default.acme]
  tlsStore = "test"
  storage = "acme.json"
  caServer = "https://acme-staging-v02.api.letsencrypt.org/directory"
  email = "test@test.com"
  keyType = "RSA4096"
  acmeLogging = true

  [certificatesResolvers.default.acme.DNSchallenge]
```

### Dynamic

```toml
[http.routers.router1]
  service = "service2"
  rule = "Host(`goo.snitest.com`)"
  [http.routers.router1.tls]
#    store = "test"
    certResolver = "default"
    [[http.routers.router1.tls.domain]]
    main = "*.snitest.com"
#    SANs = ["foo.com", "fii.com", "fuu.com"]
#    certResolver = "default"
```

### Related use cases

- Host + wildcard
- HostRegex + wildcard
- no Host + wildcard
- Host + domain
- no Host + domain
- HostRegex + domain

## Case: wildcards

- Hosts + wildcards

### Static

```toml
[certificatesResolvers.default.acme]
  tlsStore = "test"
  storage = "acme.json"
  caServer = "https://acme-staging-v02.api.letsencrypt.org/directory"
  email = "test@test.com"
  keyType = "RSA4096"
  acmeLogging = true

  [certificatesResolvers.default.acme.DNSChallenge]
```

### Dynamic

```toml
[http.routers.router1]
  service = "service2"
  rule = "Host(`goo.snitest.com`) || Host(`goo.foobar.org`)"
  [http.routers.router1.tls]
    store = "default"
    certResolver = "default"
    [[http.routers.router1.tls.domain]]
      main = "*.snitest.com"
#      certResolver = "default"
    [[http.routers.router1.tls.domain]]
      main = "*.foobar.org"
#      certResolver = "default"
```

## Case: Multiple challenges LE

- multiple challenges => type choice made by Let's encrypt (ex: no wildcard => TLS ALPN, wildcard => DNS)

No fallback.

### Static

```toml
[certificatesResolvers.default.acme]
  tlsStore = "test"
  storage = "acme.json"
  caServer = "https://acme-staging-v02.api.letsencrypt.org/directory"
  email = "test@test.com"
  keyType = "RSA4096"
  acmeLogging = true

  [certificatesResolvers.default.acme.TLSchallenge]

  [certificatesResolvers.default.acme.DNSChallenge]
```

### Dynamic

```toml
[http.routers.router1]
  service = "service2"
  rule = "Host(`goo.snitest.com`)"
  [http.routers.router1.tls]
#    store = "test"
    certResolver = "default"
```

## Case: Multiple challenges User

- multiple challenges => choice of type made by the user (domain1 => TLS ALPN, domain2 => HTTP)
- multiple challenges => choice of type and DNS made by the user (domain1 => DNS1, domain2 => DNS2)

No multi-account on the same DNS provider.
No fallback.

### Static

```toml
[certificatesResolvers.default.acme]
  tlsStore = "test"
  storage = "acme.json"
  caServer = "https://acme-staging-v02.api.letsencrypt.org/directory"
  email = "test@test.com"
  keyType = "RSA4096"
  acmeLogging = true

  [certificatesResolvers.default.acme.TLSchallenge]

[certificatesResolvers.cloudflare.acme]
  tlsStore = "test"
  storage = "acme.json"
  caServer = "https://acme-staging-v02.api.letsencrypt.org/directory"
  email = "test@test.com"
  keyType = "RSA4096"
  acmeLogging = true

  [certificatesResolvers.cloudflare.acme.DNSChallenge]
  provider = "cloudflare"
```

### Dynamic

```toml
[http.routers.router1]
  service = "service2"
  rule = "Host(`goo.snitest.com`)"
  [http.routers.router1.tls]
#   store = "test"
   certResolver = "default"

[http.routers.router2]
  service = "service2"
  rule = "Host(`goo.snitest.org`)"
  [http.routers.router2.tls]
# If "store" different from the store configured on the challenge => log.Warn and we use that of the challenge.
#    store = "test"
   certResolver = "cloudflare"

[http.routers.router3]
  service = "service2"
  rule = "Host(`goo.snitest.com`) || Host(`goo.foobar.org`) || Host(`goo.foobar.net`)"
  [http.routers.router3.tls]
#    store = "test"
    certResolver = "default"
    [[http.routers.router3.tls.domain]]
    main = "*.snitest.com"
#    certResolver = "default"
    [[http.routers.router3.tls.domain]]
    main = "*.foobar.org"
    certResolver = "cloudflare"
```

## Case: Multiple certificates providers

- Multiple certificates providers (acme, vault, ...)

No fallback.

### Static

```toml
[certificatesResolvers.lego.acme]
  tlsStore = "test"
  storage = "acme.json"
  caServer = "https://acme-staging-v02.api.letsencrypt.org/directory"
  email = "test@test.com"
  keyType = "RSA4096"
  acmeLogging = true

  [certificatesResolvers.lego.acme.TLSchallenge]

[certificatesResolvers.hashi.vault]
  vaultServer = ""
  storage = ""
  pkiPath = ""
  role = ""
```

### Dynamic

```toml
[http.routers.router1]
  service = "service2"
  rule = "Host(`goo.snitest.com`)"
  [http.routers.router1.tls]
# If "store" different from the store configured on the challenge => log.Warn and we use that of the challenge.
#    store = "test"
   certResolver = "lego"

[http.routers.router2]
  service = "service2"
  rule = "Host(`goo.snitest.org`)"
  [http.routers.router2.tls]
# If "store" different from the store configured on the challenge => log.Warn and we use that of the challenge.
#    store = "test"
   certResolver = "hashi"
```

## Case: Disable Host rule detection

- Host + `onHostRule`: false
- no explicit option `onHostRule`

If no named resolver == Host rule detection disabled

```toml
[http.routers.router1]
  service = "service2"
  rule = "Host(`goo.snitest.com`) || Host(`goo.snitest.org`)"
  [http.routers.router1.tls]
#    store = "test"
```

**Comment by EugenMayer** [comment]
Thank you for working on this one and already targeting this for 2.0 - this seems to be a really huge release all over the place. 

I am tensed to see how you manage to come up with release notes for the final, which do not look like a assay :)

**Comment by emilevauge** [comment]
>which do not look like a assay
It will, definitively 😁

**Comment by pjeby** [comment]
Does this cover multiple DNS provider credentials, e.g. domains belonging to multiple Cloudflare accounts, and thus needing different keys?

**Comment by ldez** [comment]
> ## Case: Multiple challenges User
>
> No multi-account on the same DNS provider.

For now, lego (the lib we use to manage acme) doesn't support that.

**Comment by traefiker** [comment]
Closed by #5116.
